### PR TITLE
Push chainspec type graph down into casper types

### PR DESCRIPTION
**Pull request by Fraser999**
This is #4021 rebased on `feat-2.0`.

The majority of the commits are a result of merging `dev`.  Only the commits onwards from [ea1259c](https://github.com/casper-network/casper-node/pull/4028/commits/ea1259cc0b08baf5dca3cf94911d288c835c4ae4) relate to moving the chainspec and digest types down to `casper-types`.

### CODE REVIEW

File: types/src/chainspec.rs [hunk 1]
```diff
@@ -78,13 +103,18 @@ pub struct Chainspec {
 impl Chainspec {
     /// Serializes `self` and hashes the resulting bytes.
     pub fn hash(&self) -> Digest {
-        let serialized_chainspec = self.to_bytes().unwrap_or_else(|error| {
-            error!(%error, "failed to serialize chainspec");
-            vec![]
-        });
+        let serialized_chainspec = self.to_bytes().unwrap_or_else(|_| vec![]);
```
Comment (fizyk20) [inline types/src/chainspec.rs:106]: Is it okay to just quietly return the hash of an empty vec without leaving any sign that this happened? It makes me a bit uncomfortable :grimacing:

File: types/src/chainspec.rs [hunk 2]
```diff
@@ -78,13 +103,18 @@ pub struct Chainspec {
...
         Digest::hash(serialized_chainspec)
     }
 
+    /// Serializes `self` and hashes the resulting bytes, if able.
+    pub fn maybe_hash(&self) -> Result<Digest, String> {
```
Comment (fizyk20) [inline types/src/chainspec.rs:111]: Nit: how about `try_hash`? I think that's the usual convention for functions returning `Result`s, isn't it?

**Review by EdHastingsCasperAssociation — APPROVED** [review]

**Comment by fizyk20** [comment]
I haven't managed to finish the review before this got merged; I actually tried to cancel the comments I made above, but I must have clicked something incorrectly, because they got posted instead :sweat_smile: Feel free to ignore.

**Pull request merged**